pkg/metautil/metafile.go

```diff
@@ -0,0 +1,649 @@
+	storage           storage.ExternalStorage
+	metafileSizeLimit int
+	// a flag to control whether we generate v1 or v2 meta.
+	useV2Meta  bool
+	backupMeta *backuppb.BackupMeta
+	// used to generate MetaFile name.
+	metafileSizes  map[string]int
+	metafileSeqNum map[string]int
+	metafiles      *sizedMetaFile
+	// the start time of StartWriteMetas
+	// it's use to calculate the time costs.
+	start time.Time
+	// wg waits StartWriterMetas exits
+	wg sync.WaitGroup
+	// internal item channel
+	metasCh chan interface{}
+	errCh   chan error
+
+	// records the total item of in one write meta job.
+	flushedItemNum int
+}
+
+// NewMetaWriter creates MetaWriter.
+func NewMetaWriter(storage storage.ExternalStorage, metafileSizeLimit int, useV2Meta bool) *MetaWriter {
+	return &MetaWriter{
+		start:             time.Now(),
+		storage:           storage,
+		metafileSizeLimit: metafileSizeLimit,
+		useV2Meta:         useV2Meta,
+		// keep the compatibility for old backupmeta.Ddls
+		// old version: Ddls, _ := json.Marshal(make([]*model.Job, 0))
+		backupMeta:     &backuppb.BackupMeta{Ddls: []byte("[]")},
+		metafileSizes:  make(map[string]int),
+		metafiles:      NewSizedMetaFile(metafileSizeLimit),
+		metafileSeqNum: make(map[string]int),
+	}
+}
+
+func (writer *MetaWriter) reset() {
+	writer.metasCh = make(chan interface{}, MaxBatchSize)
+	writer.errCh = make(chan error)
+
+	// reset flushedItemNum for next meta.
+	writer.flushedItemNum = 0
+}
+
+// Update updates some property of backupmeta.
+func (writer *MetaWriter) Update(f func(m *backuppb.BackupMeta)) {
+	f(writer.backupMeta)
+}
+
+// Send sends the item to buffer.
+func (writer *MetaWriter) Send(m interface{}, op AppendOp) error {
+	select {
+	case writer.metasCh <- m:
+	// receive an error from StartWriteMetasAsync
+	case err := <-writer.errCh:
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (writer *MetaWriter) close() {
+	close(writer.metasCh)
+}
+
+// StartWriteMetasAsync writes four kind of meta into backupmeta.
+// 1. file
+// 2. schema
+// 3. ddl
+// 4. rawRange( raw kv )
+// when useBackupMetaV2 enabled, it will generate multi-level index backupmetav2.
+// else it will generate backupmeta as before for compatibility.
+// User should call FinishWriteMetas after StartWriterMetasAsync.
+func (writer *MetaWriter) StartWriteMetasAsync(ctx context.Context, op AppendOp) {
+	writer.reset()
+	writer.start = time.Now()
+	go func() {
+		writer.wg.Add(1)
+		defer func() {
+			writer.wg.Done()
+			// close errCh after metaCh closed
+			close(writer.errCh)
+		}()
+		for {
+			select {
+			case <-ctx.Done():
+				log.Info("exit write metas by context done")
+				return
+			case meta, ok := <-writer.metasCh:
+				if !ok {
+					log.Info("write metas finished", zap.String("type", op.name()))
+					return
+				}
+				needFlush := writer.metafiles.append(meta, op)
+				if writer.useV2Meta && needFlush {
+					err := writer.flushMetasV2(ctx, op)
+					if err != nil {
+						writer.errCh <- err
```

Comment:
I think it's better to use a select to send this err because if there are multi routines trying to send err, only the first err will be received and other err will block

Reply:
for current structure there is only one routine in StartWriteMetasAsync.